internal/agent/configcontroller/config_controller.go

```diff
@@ -0,0 +1,262 @@
+	updateDelay = 5 * time.Second
+)
+
+type ConfigController struct {
+	caFilePath           string
+	device               *device.Device
+	deviceWriter         *device.Writer
+	deviceStatusExporter export.DeviceStatus
+	deviceObserver       *observe.Device
+	queue                workqueue.RateLimitingInterface
+
+	enrollmentClient        *client.Enrollment
+	enrollmentVerifyBackoff wait.Backoff
+	enrollmentEndpoint      string
+
+	managementClient       *client.Management
+	managementEndpoint     string
+	managementCertFilePath string
+	managementKeyFilePath  string
+
+	// The device fingerprint
+	enrollmentCSR []byte
+	// The log prefix used for testing
+	logPrefix string
+	// The directory to write the certificate to
+	certDir string
+}
+
+func New(
+	device *device.Device,
+	enrollmentClient *client.Enrollment,
+	enrollmentEndpoint string,
+	managementEndpoint string,
+	caFilePath string,
+	managementCertFilePath string,
+	managementKeyFilePath string,
+	deviceWriter *device.Writer,
+	deviceStatusExporter export.DeviceStatus,
+	deviceObserver *observe.Device,
+	enrollmentCSR []byte,
+	logPrefix string,
+) *ConfigController {
+
+	enrollmentVerifyBackoff := wait.Backoff{
+		Cap:      3 * time.Minute,
+		Duration: 10 * time.Second,
+		Factor:   1.5,
+		Steps:    24,
+	}
+
+	c := &ConfigController{
+		enrollmentClient:        enrollmentClient,
+		enrollmentVerifyBackoff: enrollmentVerifyBackoff,
+		enrollmentEndpoint:      enrollmentEndpoint,
+		device:                  device,
+		deviceWriter:            deviceWriter,
+		deviceStatusExporter:    deviceStatusExporter,
+		deviceObserver:          deviceObserver,
+		caFilePath:              caFilePath,
+		managementEndpoint:      managementEndpoint,
+		managementCertFilePath:  managementCertFilePath,
+		managementKeyFilePath:   managementKeyFilePath,
+		enrollmentCSR:           enrollmentCSR,
+		logPrefix:               logPrefix,
+	}
+
+	c.queue = workqueue.NewNamedRateLimitingQueue(workqueue.NewMaxOfRateLimiter(
+		&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(updateDelay), 1)},
+		workqueue.NewItemExponentialFailureRateLimiter(1*time.Second, maxUpdateBackoff)), "device-config")
+
+	return c
+}
+
+func (c *ConfigController) sync(ctx context.Context, device *v1alpha1.Device) error {
+	deviceStatus := c.deviceStatusExporter.Get(ctx)
+	// ensure the device is enrolled
+	if err := c.ensureDeviceEnrollment(ctx, device); err != nil {
+		klog.Errorf("%s enrollment did not succeed: %v", c.logPrefix, err)
+		return err
+	}
+
+	if !c.deviceObserver.HasSynced(ctx) {
+		return fmt.Errorf("device observer has not synced")
+	}
+
+	var conditions []v1alpha1.DeviceCondition
+	// post enrollment update status
+	deviceCondition := v1alpha1.DeviceCondition{
+		Type:   "Enrolled",
+		Status: v1alpha1.True,
+	}
+	conditions = append(conditions, deviceCondition)
+	deviceStatus.Conditions = &conditions
+	_, updateErr := c.managementClient.UpdateDeviceStatus(ctx, *device.Metadata.Name, deviceStatus)
```

Comment:
Is this message sent on every call to sync(), even if the device is already enrolled?

Reply:
makes sense to batch this.